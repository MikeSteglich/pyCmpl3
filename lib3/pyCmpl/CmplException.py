#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

#from .CmplMsg import *
#from html.parser import HTMLParser
import html

from io import StringIO

#*************** CmplException ***********************************
class CmplException(Exception):
	
	def __init__(self, msg, msgList=None):
		self.__msgStr=StringIO()

		if type(msg) == bytes:
			msg = msg.decode("utf-8")

		if not "pyCmpl error:" in msg:
			self.__msgStr.write("pyCmpl error: " +msg)
		else:	
			self.__msgStr.write(msg)
		
		if msgList!=None:
			for m in msgList:
				self.__msgStr.write("\n" + m.type + " in module <" + m.module + "> at location <" + m.location + "> " + html.unescape(m.description))
	
	def __str__(self):
		return "pyCmpl error:"
	
	def __del__(self):
		self.__msgStr.close()
	
	@property
	def msg(self):
		return self.__msgStr.getvalue()
#*************** end CmplException *******************************


		
		
		
		

		