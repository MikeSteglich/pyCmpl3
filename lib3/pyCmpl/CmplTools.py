#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 

import re
import math


#*************** CmplTools ***************************************
class CmplTools(object):

	#*********** stringContainsWhitespace **********
	@staticmethod
	def stringContainsWhitespace(string):
		regex = re.compile('[\t\n\x0b\x0c\r ]+')
		if regex.search(string):
			return True
		else:
			return False
	#*********** end stringContainsWhitespace ******
	
	#*********** strIsNumber ***********************
	@staticmethod
	def strIsNumber(str):
		r = re.compile('[^0-9.]+')
		if  r.search(str):
			return False
		else:
			return True
	#*********** end strIsNumber *******************
	
	#*********** strIsValidIndex ***********************
	@staticmethod
	def strIsValidIndex(str):
		r = re.compile('[^0-9a-zA-Z\-_.]+')
		if  r.search(str):
			return False
		else:
			return True
	#*********** end strIsValidIndex *******************
	
	
	
	#*********** strContains ***********************
	@staticmethod
	def strContains(str0, str1):
		return str1 in str0
	#*********** end strContains *******************
	
	@staticmethod
	def haversine(lat1, lon1, lat2, lon2):
		'''http://rosettacode.org/wiki/Haversine_formula#Python'''
	     
		R = 6372.8 # Earth radius in kilometers
	     
		dLat = math.radians(lat2 - lat1)
		dLon = math.radians(lon2 - lon1)
		lat1 = math.radians(lat1)
		lat2 = math.radians(lat2)
	     
		a = math.sin(dLat/2)**2 + math.cos(lat1)*math.cos(lat2)*math.sin(dLon/2)**2
		c = 2*math.asin(math.sqrt(a))
	     
		return R * c
	

#*************** end CmplTools **********************************


		
		
		

		