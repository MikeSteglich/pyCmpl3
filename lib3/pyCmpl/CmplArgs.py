#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************


import os
import sys

from .CmplDefs import *
from .CmplException import *

#*************** CmplArgs ***********************************
class CmplArgs:
	
	def __init__(self, optFile=None):

		self.__optFile = optFile
		
		self.__cmplFile=""
		self.__modelName=""	

		self.__url=""
		self.__maxTries = 10
		self.__maxTime = 300
		
		self.__externalFiles = []

		self.__xlsDataFile = ""
		self.__messageFile=""
		self.__matrixFile=""
		self.__solFile=""
		self.__solAsciiFile=""
		self.__solCsvFile=""

		self.__isFileOut=False

		self.__runMode = CMPL_LOCAL

		self.__isSilent = False

		self.__solver = "cbc"

	#******** getter **********************************************		
	@property
	def optFile(self):
		return self.__optFile

	@property
	def optAlias(self):
		return os.path.splitext(self.__optFile)[0]

	@property
	def cmplFile(self):
		return self.__cmplFile
	
	@property
	def baseName(self):
		return self.__modelName

	@property
	def url(self):
		return self.__url

	@property
	def maxTries(self):
		return self.__maxTries

	@property
	def maxTime(self):
		return self.__maxTime

	@property
	def runMode(self):
		return self.__runMode

	def setRunMode(self, mode):
		if mode in (CMPL_LOCAL, CMPL_REMOTE_SOLVE , CMPL_REMOTE_SEND , CMPL_REMOTE_KNOCK , CMPL_REMOTE_RETRIEVE , CMPL_REMOTE_CANCEL ,PYCMPL):
			self.__runMode = mode
		else:
			raise CmplException("Wrong running mode ")
	
	@property
	def msgFile(self):
		return self.__messageFile

	@property
	def xlsDataFile(self):
		return self.__xlsDataFile

	@property
	def externalFiles(self):
		return self.__externalFiles

	@property
	def msgFile(self):
		return self.__messageFile

	@property
	def solFile(self):
		return self.__solFile

	@property
	def solAsciiFile(self):
		return self.__solAsciiFile

	@property
	def solCsvFile(self):
		return self.__solCsvFile
	
	@property
	def isFileOut(self):
		return self.__isFileOut

	@property
	def isSilent(self):
		return self.__isSilent

	@property
	def solver(self):
		return self.__solver

	#******** getter **********************************************


	#******** parseOptFile **********************************************
	def parseOptFile(self):
		if not self.__optFile:
			raise CmplException("Cannot parse option file, because it is not defined")

		nrOfXlsDataFiles=0
		
		try:
			f = open(self.__optFile, "r")
			opts = f.read()
			f.close()

			if not opts:
				raise CmplException("Cannot read options form opt file")

			lines = opts.split(os.linesep)

			for line in lines:
				line=line.strip()
				
				if line.startswith("#"):
					continue

				if not line:
					continue

				optList=line.split(";")

				key = optList[0].upper()
				if optList[9]=="0":
					val=""
				else:
					val = optList[10]

				if key=="I":
					self.__cmplFile=val[1:-1]
					self.__modelName = os.path.splitext(self.__cmplFile)[0]

				elif key=="XLSDATA":
					if not val or val[1:-1]==":":
						self.__xlsDataFile = self.__modelName+".xdat"
					else:	
						self.__xlsDataFile=val[1:-1]
					nrOfXlsDataFiles+=1

				elif key=="DATA":
					tmpName=""
					if not val or val[1:-1]==":":
						tmpName = self.__modelName+".cdat"
					else:	
						tmpName=val[1:-1]
					self.__externalFiles.append(tmpName)

				elif key=="INCLUDE":
					self.__externalFiles.append(val[1:-1])
					
				elif key=="URL":
					self.__url = val[1:-1]
					self.__runMode = CMPL_REMOTE_SOLVE

				elif key=="SEND":
					self.__runMode = CMPL_REMOTE_SEND

				elif key=="KNOCK":
					self.__runMode = CMPL_REMOTE_KNOCK

				elif key=="RETRIEVE":
					self.__runMode = CMPL_REMOTE_RETRIEVE

				elif key=="CANCEL":
					self.__runMode = CMPL_REMOTE_CANCEL

				elif key=="MAXTRIES":
					self.__maxTries = int(val[1:-1])

				elif key=="MAXTIME":
					self.__maxTime  = int(val[1:-1])

				elif key=="CMSG":
					if not val:
						self.__messageFile=self.__modelName+".cmsg"
					else:	
						self.__messageFile=val[1:-1]

				elif key=="SOLUTION":
					if not val:
						self.__solFile=self.__modelName+".csol"
					else:	
						self.__solFile=val[1:-1]
			
				elif key=="SOLUTIONASCII":
					if not val:
						self.__solAsciiFile=self.__modelName+".sol"
					else:	
						self.__solAsciiFile=val[1:-1]
				
				elif key=="SOLUTIONCSV":
					if not val:
						self.__solCsvFile=self.__modelName+".csv"
					else:	
						self.__solCsvFile=val[1:-1]

				elif key in ("M","FM","P","MATRIX"):
					self.__isFileOut=True
				
				elif key=="SILENT":
					self.__isSilent=True
				
				elif key=="SOLVER":
					self.__solver=val[1:-1]

			if nrOfXlsDataFiles>1:
				raise CmplException("Only one CmplXlsData file per Cmpl model is allowed")
		
		except:
			raise CmplException("Something went wrong while parsing Cmpl arguments: "+str(sys.exc_info()[1]))
	#******** parseOptFile **********************************************



#*************** end CmplArgs *******************************


		
        
        
        

        
