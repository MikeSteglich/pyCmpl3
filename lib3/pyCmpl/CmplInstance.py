#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

import os
import xml.dom.minidom as dom
import random
import io
from xml.sax.saxutils import unescape, escape

from .CmplException import *

#*************** CmplInstance ***********************************
class CmplInstance(object):

	#*********** constructor **********
	def __init__(self):
		
		self.__cmplFile = ""
		self.__cmplAlias = ""
		self.__cmplContent=""
		self.__optionsList = []
		self.__dataString = ""
		self.__cmplFileList = {}
		self.__instStr = io.StringIO()
		self.__jobId = ""
		self.__cmplName = ""
		self.__args=None
		self.__withPreComp=False
		
	#*********** end constructor ******
	
	#*********** destructor ***********
	def __del__(self):
		self.__instStr.close()
	#*********** end destructor *******
		
	# getter **************************	
	@property
	def jobId(self):
		return self.__jobId
		
	@property
	def options(self):
		return self.__optionsList

	@property
	def withPreComp(self):
		return self.__withPreComp
	# end getter ***********************	
	

	def __readAsciiFile(self, fName):
		fContent=""
		try:
			f = open(fName, "r")
			fContent = f.read()
			f.close()		
			return fContent
		except IOError as e:
			raise CmplException("IO error for file <"+fName+"> : "+str(e))

	def __isFullPath(self, fName):
		fullPath=False
		if fName.startswith(os.sep) or fName[1:2] == ":":
			fullPath = True 
		return fullPath

	def __getExternalFiles(self):

		isFullPath=self.__isFullPath(self.__cmplFile)
				
		self.__cmplAlias=os.path.splitext(os.path.basename(self.__cmplFile))[0] 

		fAlias=self.__cmplAlias+".optcmpl"
		self.__cmplFileList[fAlias]=self.__readAsciiFile(self.__args.optAlias+".optcmpl")
		self.__optionsList.append("-i-opt "+fAlias)

		fAlias=self.__cmplAlias+".precmpl"
		self.__cmplFileList[fAlias]=self.__readAsciiFile(self.__args.optAlias+".precmpl")
		self.__optionsList.append("-i-pre "+fAlias)

		fAlias=self.__cmplAlias+".extdata"
		self.__cmplFileList[fAlias]=self.__readAsciiFile(self.__args.optAlias+".extdata")
		self.__optionsList.append("-i-extern "+fAlias)

		if self.__args.xlsDataFile:
			fAlias=self.__cmplAlias+".cdat"

			if isFullPath and not self.__isFullPath(self.__args.xlsDataFile):
				self.__optionsList.append("-filealias "+os.path.dirname(self.__cmplFile)+os.sep+self.__args.xlsDataFile+"="+fAlias)
			else:
				self.__optionsList.append("-filealias "+self.__args.xlsDataFile+"="+fAlias)
			self.__cmplFileList[fAlias] = self.__dataString
					
		for fName in self.__args.externalFiles:
			fAlias=os.path.basename(fName)
			if fAlias[-1]==":":
				fAlias=fAlias[:-1]
				fName=fName[:-1]

			if fName!=fAlias: 
				if isFullPath and not self.__isFullPath(fName):
					self.__optionsList.append("-filealias "+os.path.dirname(self.__cmplFile)+os.sep+fName+"="+fAlias)
				else:
					self.__optionsList.append("-filealias "+fName+"="+fAlias)

			if fAlias==self.__cmplAlias+ ".cdat":
				if not self.__dataString:
					self.__cmplFileList[fAlias] = self.__readAsciiFile(fName)	
				else:
					self.__cmplFileList[fAlias] = self.__dataString
			else:
				self.__optionsList.append("-filealias "+fName+"="+fAlias)
				self.__cmplFileList[fAlias] = self.__readAsciiFile(fName)	

		if isFullPath:
			self.__optionsList.append("-basename "+self.__cmplAlias)
			
		
	#*********** cmplInstanceStr **********
	def cmplInstanceStr(self, cmplFileName, cmplArgs, optList, dataString, jobId):

		self.__args=cmplArgs
		self.__cmplFile = cmplFileName

		self.__optionsList = optList
		self.__jobId = jobId
		self.__dataString = dataString

		self.__getExternalFiles()
		
		try:
			self.__instStr.write("<?xml version = \"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n")
			self.__instStr.write("<CmplInstance version=\"2.0\">\n")
			self.__instStr.write("<general>\n")
			self.__instStr.write("<name>"+os.path.basename(self.__cmplFile)+"</name>\n")
			self.__instStr.write("<jobId>"+self.__jobId+"</jobId>\n")
			self.__instStr.write("<preComp>no</preComp>\n")
			self.__instStr.write("</general>\n")
			
			if len(self.__optionsList) > 0:
				self.__instStr.write("<options>\n")
				for opt in self.__optionsList:
					self.__instStr.write("<opt>"+opt+"</opt>\n")
				self.__instStr.write("</options>\n")
			
			self.__instStr.write("<problemFiles>\n")
			
			for d in self.__cmplFileList:
				self.__instStr.write("<file name=\""+ d + "\">\n")
				self.__instStr.write(escape(self.__cmplFileList[d]) )
				self.__instStr.write("\n")	
				self.__instStr.write("</file>\n")
			self.__instStr.write("</problemFiles>\n")	
			self.__instStr.write("</CmplInstance>\n")	

		except IOError as e:
			raise CmplException("IO error : "+str(e))
		
		return self.__instStr.getvalue()
		
	#*********** end cmplInstanceStr ******	

		
	#*********** writeCmplInstance **********
	def writeCmplInstance(self, folder, instStr):	
		if os.path.exists(folder) == False:
			raise CmplException("Path <"+self.__cmplServerPath+"> doesn't exist.")
			
		instDom = dom.parseString(instStr)

		if instDom.firstChild.nodeName!="CmplInstance":
			raise CmplException("Cant't read cmplInstance file - wrong file type!")
			
		for entry in instDom.firstChild.childNodes: 
				if entry.nodeName == "general":
					for entry1 in entry.childNodes:
						if entry1.nodeName == "name":
							self.__cmplName = unescape( entry1.firstChild.data.strip() , entities={ "&apos;":"'"} )
							continue
						if entry1.nodeName == "jobId":
							self.__jobId = entry1.firstChild.data.strip()
							continue
						if entry1.nodeName == "preComp":
							if entry1.firstChild.data.strip()=="no":
								self.__withPreComp=False
							elif entry1.firstChild.data.strip()=="yes":
								self.__withPreComp=True
							continue
							
				if entry.nodeName == "options":
					for entry1 in entry.childNodes: 
						if entry1.nodeName == "opt":
							self.__optionsList.append(entry1.firstChild.data.strip())
							
				if entry.nodeName == "problemFiles":
					for entry1 in entry.childNodes: 
						if entry1.nodeName == "file":
							tmpName = folder+self.__jobId+os.sep+entry1.getAttribute("name")
							tmpContent = unescape(entry1.firstChild.data.strip() , entities={ "&apos;":"'"})
							try:
								f = open(tmpName, 'w')
								f.write(tmpContent)		
								f.close()
							except IOError as e:
								raise CmplException("IO error for file " + tmpName + ": "+e)
	#*********** end writeCmplInstance *******

#*************** end CmplInstance  ******************************

		