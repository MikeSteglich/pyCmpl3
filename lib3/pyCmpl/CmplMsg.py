#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

import os
import re
import io 

from xml.sax.saxutils import unescape

from .CmplDefs import *
from .CmplException import *


#*************** CmplMsg *****************************************
class CmplMsg(object):
	def __init__(self):
		self.__type = ""
		self.__module = ""
		self.__location = ""
		#self.__line = ""
		self.__description = ""
		
	# getter and setter 	
	@property
	def type(self):
		return self.__type
	
	def setType(self, type):
		self.__type=type
		
	@property
	def location(self):
		return self.__location
		
	def setLocation(self, file):
		self.__location=file

	@property
	def module(self):
		return self.__module

	def setModule(self,mod):
		self.__module=mod
		
	@property
	def description(self):
		return self.__description
		
	def setDescription(self, msg):
		self.__description=msg
	#end getter and setter
	
			
#*************** end CmplMsg *************************************


#*************** CmplMsg *****************************************
class CmplMessages(object):	

	def __init__(self, msgFile=None):
		self.__cmplStatus = ""
		self.__instance = ""
		self.__cmplVersion = ""
		self.__cmplMessage = ""
		self.__msgFile = msgFile
		self.__cmplMessageList = []
		self.__nrOfMessages = 0
		
	#*********** cmplStatus ********	
	@property
	def cmplStatus(self):
		if self.__cmplStatus=="normal":
			return CMPL_OK
		elif self.__cmplStatus=="warning":
			return CMPL_WARNINGS
		elif self.__cmplStatus=="error":
			return CMPL_FAILED
		else:
			return CMPL_UNKNOWN
	#*********** end cmplStatus *****	
	
	#*********** cmplMessageList ****	
	@property
	def cmplMessageList(self):
		return self.__cmplMessageList
	#******* end cmplMessageList ****
	
	#*********** cmplMessage ********	
	@property
	def cmplMessage(self):
		return self.__cmplMessage
	#******* end cmplMessage ********
	
		
	#*********** readCmplMessages ********	
	def readCmplMessages(self, msgStr=None):
	
		if msgStr == None and self.__msgFile == None:
			raise CmplException("Neither cmplMessageFile nor cmplMessageString defined")
		
		if 	self.__msgFile != None:
			if not os.path.isfile(self.__msgFile):
				raise CmplException("CMPL failed."  )
	
			try:	
				f = open(self.__msgFile, "r")
				msgStr = f.read()
				f.close()
			except IOError:
				raise CmplException("IO error for CmplMessage file ")	
				
		lines = io.StringIO(msgStr) 
		
		lineNr=1
		generalSection=False
		msgSection=False
		x = CmplMsg()
		
		for line in lines:
			if lineNr == 1:
				if not "<?xml version" in line: 
					raise CmplException("File " + self.__msgFile + " - is not a XML file!")
				else:
					lineNr += 1
					continue
			if lineNr == 2:
				if not "<CmplMessages" in line: 
					raise CmplException("Cant't read cmplMessage file " + self._cmplMsgFile + " - wrong file type!")
				else:
					lineNr += 1
					continue
	
			if "<general>" in line:
				generalSection=True
				msgSection=False
				continue
			if "</general>" in line:
				generalSection=False
				continue
			if "<messages" in line:
				generalSection=False
				msgSection=True
				continue
			if "</messages" in line:
				msgSection=False
				continue
				
			if generalSection:
				if "<generalStatus" in line:
					self.__cmplStatus = re.findall("<generalStatus>([^\"]*)</generalStatus>", line)[0]
					continue
				if "<instanceName" in line:
					self.__instance = re.findall("<instanceName>([^\"]*)</instanceName>", line)[0]
					path,name=os.path.split(self.__instance)
					self.__instance=name

				if "<message" in line:
					self.__cmplMessage = re.findall("<message>([^\"]*)</message>", line)[0]
					continue
				if "<cmplVersion" in line:
					self.__cmplVersion = re.findall("<cmplVersion>([^\"]*)</cmplVersion>", line)[0]
					continue
		
			if msgSection:
				if "<message" in line:
					tmpList=re.findall("\"([^\"]*)\"", line)
					x = CmplMsg()
					x.setType(tmpList[0])

					x.setModule(tmpList[1])

					tmpLocation=""
					path,name=os.path.split(tmpList[2])

					tmpLoc=re.findall('(.*)_cmpl__(.*).cmpl(.*)',tmpList[2])
					if len(tmpLoc)>0:
						tmpLocation=tmpLoc[0][0]+".cmpl"+tmpLoc[0][0]
					else:
						tmpLocation=tmpList[2]
					
					path,name=os.path.split(tmpLocation)
					x.setLocation(name)
		
					x.setDescription( unescape(tmpList[3] , entities={ "&apos;":"'"} ) )
					self.__cmplMessageList.append(x)

		self.__nrOfMessages= len(self.__cmplMessageList)
						
							
	#*********** end readCmplMessages ****	
		
	
	#*********** writeCmplMessages ********	
	def writeCmplMessages(self, fileName):
		try:	
			f = open(fileName, "w")
			f.write("<?xml version=\"1.0\" encoding=\"UTF-8\"?>"+os.linesep)
			f.write("<CmplMessages version=\"1.1\">"+os.linesep)
			f.write("	<general>"+os.linesep)
			f.write("		<instanceName>"+self.__instance+"</instanceName>"+os.linesep)
			f.write("		<generalStatus>"+self.__cmplStatus+"</generalStatus>"+os.linesep)
			f.write("		<message>"+self.__cmplMessage+"</message>"+os.linesep)
			f.write("		<cmplVersion>"+self.__cmplVersion+"</cmplVersion>"+os.linesep)
			f.write("	</general>"+os.linesep)

			if self.__nrOfMessages>0:
				f.write("	<messages numberOfMessages=\""+str(self.__nrOfMessages)+"\">"+os.linesep)

				for m in self.__cmplMessageList:
					f.write("		<message type =\""+m.type+"\" module=\""+m.module+"\" location=\""+m.location+"\" description=\""+m.description+"\"/>"+ os.linesep)

				f.write("	</messages>"+os.linesep)
			
			f.write("</CmplMessages>"+os.linesep)

			f.close()
		except IOError:
			raise CmplException("IO error for CmplMessage file ")
	#*********** end writeCmplMessages ********	
		
		
		
		

		