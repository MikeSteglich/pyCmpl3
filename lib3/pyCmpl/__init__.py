from .CmplDefs import *
from .Cmpl import *
from .CmplException import *
from .CmplMsg import *
from .CmplTools import *
from .CmplSolution import *
from .CmplSet import *
from .CmplParameter import *
from .CmplInstance import *
from .CmplXlsData import *
from .CmplArgs import *
from .CmplSolutionReport import *


	