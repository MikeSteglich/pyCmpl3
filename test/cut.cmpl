%data : rollWidth , widths set, patterns set, orders[widths],  nbr[widths, patterns]

variables:
	 cut[patterns]: integer[0..];

objectives:
	number: sum{ j in patterns: cut[j] }->min;

constraints: 
	 {i in widths: fill[i]: sum{ j in patterns : nbr[i,j] * cut[j] } >= orders[i];} 

 
 
 
 
 
 
 
 
 
 
 
 
 
 
