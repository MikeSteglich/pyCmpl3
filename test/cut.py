#!/usr/bin/python 

from pyCmpl import *
import math

try: 
	cuttingOpt = Cmpl("cut.cmpl")
	patternGen = Cmpl("cut-pattern.cmpl")


	#cuttingOpt.setOption("-solver cplex")
	cuttingOpt.setOption("-no-remodel")
	#patternGen.setOption("-solver cplex")

	#cuttingOpt.setOutput(True)
	#patternGen.setOutput(True)

	#patternGen.debug(True)

	r = CmplParameter("rollWidth")
	r.setValues(110)
	
	w = CmplSet("widths")
	w.setValues([ 20, 45, 50, 55, 75])
	
	o = CmplParameter("orders",w)
	o.setValues([ 48, 35, 24, 10, 8 ])
	
	nPat=w.len
	p = CmplSet("patterns")
	p.setValues(1,nPat)
	
	nbr = []
	for i in range(nPat):
		nbr.append( [ 0 for j in range(nPat) ] )
	
	for i in w.values:
		pos = w.values.index(i)
		nbr[pos][pos] = int(math.floor( r.value / i ))
	
	n = CmplParameter("nbr", w, p)
	n.setValues(nbr)
	
	price = [] 
	for i in range(w.len):
		price.append(0)
	
	pr = CmplParameter("price", w)
	pr.setValues(price)
	
	cuttingOpt.setSets(w,p)
	cuttingOpt.setParameters(r, o, n)
	
	patternGen.setSets(w)
	patternGen.setParameters(r,pr)

	ri = cuttingOpt.setOption("-int-relax")
	
	while True:
		cuttingOpt.solve() 
		cuttingOpt.conByName()
		
		for i in w.values:
			pos = w.values.index(i)
			price[pos] = cuttingOpt.fill[i].marginal
	
		pr.setValues(price)
					
		patternGen.solve()
		patternGen.varByName()
			
		if (1-patternGen.solution.value) < -0.00001:
			nPat = nPat + 1
			p.setValues(1,nPat)
			for i in w.values:
				pos = w.values.index(i)
				nbr[pos].append(patternGen.use[i].activity)
			n.setValues(nbr)
		else:
			break
			
	cuttingOpt.delOption(ri)
	
	cuttingOpt.solve() 	
	cuttingOpt.varByName()
	
	print("Objective value: " , cuttingOpt.solution.value , "\n")
	print("Pattern:")

	vStr="   | "
	for j in p.values:
		vStr+= " %d " % j 
	print(vStr)
	
	print("----------------------------")
	for i in range(len(w.values)):
		vStr="%2d | " % w.values[i]
		for j in p.values:
			vStr += " %d " % nbr[i][j-1] 
		print(vStr)
	print("\n")
	
	for j in p.values:
		if cuttingOpt.cut[j].activity>0:
			print("%2d pieces of pattern: %d" % (cuttingOpt.cut[j].activity, j))
			for i in range(len(w.values)):
				print("   width ", w.values[i] , " - " , nbr[i][j-1]) 
	
except CmplException as e:
	print(e.msg)
	
	