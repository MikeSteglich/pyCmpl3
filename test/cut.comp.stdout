Objective value:  47.0 

Pattern:
   |  1  2  3  4  5  6  7  8 
----------------------------
20 |  5  0  0  0  0  1  1  3 
45 |  0  2  0  0  0  0  2  0 
50 |  0  0  2  0  0  0  0  1 
55 |  0  0  0  2  0  0  0  0 
75 |  0  0  0  0  1  1  0  0 


 1 pieces of pattern: 1
   width  20  -  5
   width  45  -  0
   width  50  -  0
   width  55  -  0
   width  75  -  0
 9 pieces of pattern: 3
   width  20  -  0
   width  45  -  0
   width  50  -  2
   width  55  -  0
   width  75  -  0
 5 pieces of pattern: 4
   width  20  -  0
   width  45  -  0
   width  50  -  0
   width  55  -  2
   width  75  -  0
 8 pieces of pattern: 6
   width  20  -  1
   width  45  -  0
   width  50  -  0
   width  55  -  0
   width  75  -  1
18 pieces of pattern: 7
   width  20  -  1
   width  45  -  2
   width  50  -  0
   width  55  -  0
   width  75  -  0
 6 pieces of pattern: 8
   width  20  -  3
   width  45  -  0
   width  50  -  1
   width  55  -  0
   width  75  -  0
